specutils/analysis/width.py

```diff
@@ -0,0 +1,212 @@
+    point closest to half max on either side of the maximum, and
+    measures the distance between them. This will be calculated over the
+    regions, if they are specified. 
+
+    Parameters
+    ----------
+    spectrum : `~specutils.spectra.spectrum1d.Spectrum1D`
+        The spectrum object over which the width will be calculated.
+
+    region: `~specutils.utils.SpectralRegion` or list of `~specutils.utils.SpectralRegion`
+        Region within the spectrum to calculate the FWHM value.
+
+    Returns
+    -------
+    whm : `~astropy.units.Quantity` or list (based on region input)
+        Full width of the signal at half max
+    """
+    return _computation_wrapper(_compute_fwhm, spectrum, region)
+
+
+def _compute_gaussian_fwhm(spectrum, region=None):
+    """
+    Estimate the full width of the spectrum at half max.
+
+    Parameters
+    ----------
+    spectrum : `~specutils.spectra.spectrum1d.Spectrum1D`
+        The spectrum object overwhich the width will be calculated.
+
+    region: `~specutils.utils.SpectralRegion`
+        Region within the spectrum to calculate the FWHM value.
+
+    Returns
+    -------
+    gaussian_fwhm : `~astropy.units.Quantity` or list (based on region input)
+        Approximate full width of the signal at half max
+
+    Notes
+    -----
+    This is a helper function for the above `gaussian_fwhm()` method.
+
+    """
+
+    if region is not None:
+        calc_spectrum = region.extract(spectrum)
+    else:
+        calc_spectrum = spectrum
+
+    flux = calc_spectrum.flux
+    spectrum = calc_spectrum.spectral_axis
+
+    dx = spectrum - np.mean(spectrum)
+    fwhm = 2 * np.sqrt(np.sum((dx * dx) * flux, axis=-1) / np.sum(flux, axis=-1))
+
+    return fwhm
+
+
+def _compute_gaussian_sigma_width(spectrum, region=None):
+    """
+    Estimate the full width of the spectrum based on an approximate value of
+    sigma.
+
+    Parameters
+    ----------
+    spectrum : `~specutils.spectra.spectrum1d.Spectrum1D`
+        The spectrum object overwhich the width will be calculated.
+
+    region: `~specutils.utils.SpectralRegion`
+        Region within the spectrum to calculate the gaussian sigma width.
+
+    Returns
+    -------
+    full_width : `~astropy.units.Quantity` or list (based on region input)
+        Approximate full width of the signal
+
+    Notes
+    -----
+    This is a helper function for the above `gaussian_sigma_width()` method.
+
+    """
+
+    fwhm = _compute_gaussian_fwhm(spectrum, region)
+    sigma = fwhm * gaussian_fwhm_to_sigma
+
+    return sigma * 2
+
+
+def _arghalf(array, halfval):
+    """
+    Find index of the item closest to the half max
+    """
+    return np.abs(array - halfval).argmin()
+
+
+def _compute_single_fwhm(flux, spectrum):
+
+    argmax = np.argmax(flux)
+    halfval = flux[argmax] / 2
+
+    max_idx = len(flux) - 1
+    l_idx = _arghalf(flux[:argmax], halfval) if argmax > 0 else 0
```

Comment:
Doesn't this end up finding the *left-most* cross over point instead of the one closest to the center? (But maybe just changing it to ``.argmax()`` instead of ``.argmin()`` is the fix?)

Reply:
Yes it does, this is what I mentioned to you in person yesterday. It's quick and dirty and intended to be efficient. Using `argmin` means we find the index of the minimum difference, whereas `argmax` would return the index of the maximum difference, which is not what we want. 

The better thing to do here is probably to explicitly traverse this array out from the center and look for the minimum difference, which is not quite as efficient, but might provide better results.